atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/client/creation/CqlClientFactoryImpl.java

```diff
@@ -0,0 +1,183 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.async.client.creation;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.security.NoSuchAlgorithmException;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import javax.annotation.Nonnull;
+import javax.net.ssl.SSLContext;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.CqlSessionBuilder;
+import com.datastax.oss.driver.api.core.config.DefaultDriverOption;
+import com.datastax.oss.driver.api.core.config.DriverConfigLoader;
+import com.datastax.oss.driver.api.core.context.DriverContext;
+import com.datastax.oss.driver.api.core.session.ProgrammaticArguments;
+import com.datastax.oss.driver.internal.core.context.DefaultDriverContext;
+import com.datastax.oss.driver.internal.core.context.NettyOptions;
+import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfig;
+import com.palantir.atlasdb.cassandra.CassandraServersConfigs;
+import com.palantir.atlasdb.cassandra.CassandraServersConfigs.CqlCapableConfig;
+import com.palantir.atlasdb.keyvalue.cassandra.async.CqlClient;
+import com.palantir.atlasdb.keyvalue.cassandra.async.CqlClientImpl;
+import com.palantir.atlasdb.keyvalue.cassandra.async.ThrowingCqlClientImpl;
+import com.palantir.conjure.java.config.ssl.SslSocketFactories;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+import io.netty.channel.Channel;
+import io.netty.handler.proxy.Socks5ProxyHandler;
+
+public final class CqlClientFactoryImpl implements CqlClientFactory {
+
+    public static final CqlClientFactory DEFAULT = new CqlClientFactoryImpl();
+    private static final Logger log = LoggerFactory.getLogger(CqlClientFactoryImpl.class);
+    private static final String LOAD_BALANCING_POLICY = "DcInferringLoadBalancingPolicy";
+    private static final String COMPRESSION_PROTOCOL = "lz4";
+
+    private CqlClientFactoryImpl() {
+        // Use instance
+    }
+
+    @Override
+    public CqlClient constructClient(
+            TaggedMetricRegistry taggedMetricRegistry,
+            CassandraKeyValueServiceConfig config,
+            boolean initializeAsync) {
+        return config.servers().accept(new CassandraServersConfigs.Visitor<CqlClient>() {
+            @Override
+            public CqlClient visit(CassandraServersConfigs.DefaultConfig defaultConfig) {
+                return ThrowingCqlClientImpl.SINGLETON;
+            }
+
+            @Override
+            public CqlClient visit(CqlCapableConfig cqlCapableConfig) {
+                if (!cqlCapableConfig.validateHosts()) {
```

Comment:
Hmm, we should probably make this more descriptive e.g. `thriftHostsAndCqlHostsMatch`

Reply:
Renamed to suggested.